common/txmgr/inmemory_store.go

```diff
@@ -0,0 +1,1908 @@
+package txmgr
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math/big"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	feetypes "github.com/smartcontractkit/chainlink/v2/common/fee/types"
+	txmgrtypes "github.com/smartcontractkit/chainlink/v2/common/txmgr/types"
+	"github.com/smartcontractkit/chainlink/v2/common/types"
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/label"
+	"gopkg.in/guregu/null.v4"
+)
+
+// BIG TODO LIST
+// TODO: figure out if multiple tx attempts are actually stored in the db for each tx
+// TODO: need a way to get id for a tx attempt. since there are some methods where the persistent store creates a tx attempt and doesnt returns it
+// TODO: should txAttempt state transitions be handled by the address state manager?
+
+var (
+	// ErrInvalidChainID is returned when the chain ID is invalid
+	ErrInvalidChainID = errors.New("invalid chain ID")
+	// ErrTxnNotFound is returned when a transaction is not found
+	ErrTxnNotFound = errors.New("transaction not found")
+	// ErrExistingIdempotencyKey is returned when a transaction with the same idempotency key already exists
+	ErrExistingIdempotencyKey = errors.New("transaction with idempotency key already exists")
+	// ErrAddressNotFound is returned when an address is not found
+	ErrAddressNotFound = errors.New("address not found")
+	// ErrSequenceNotFound is returned when a sequence is not found
+	ErrSequenceNotFound = errors.New("sequence not found")
+	// ErrCouldNotGetReceipt is the error string we save if we reach our finality depth for a confirmed transaction without ever getting a receipt
+	// This most likely happened because an external wallet used the account for this nonce
+	ErrCouldNotGetReceipt = errors.New("could not get receipt")
+)
+
+type PersistentTxStore[
+	ADDR types.Hashable,
+	CHAIN_ID types.ID,
+	TX_HASH types.Hashable,
+	BLOCK_HASH types.Hashable,
+	R txmgrtypes.ChainReceipt[TX_HASH, BLOCK_HASH],
+	SEQ types.Sequence,
+	FEE feetypes.Fee,
+] interface {
+	txmgrtypes.TxStore[ADDR, CHAIN_ID, TX_HASH, BLOCK_HASH, R, SEQ, FEE]
+
+	UnstartedTransactions(limit, offset int, fromAddress ADDR, chainID CHAIN_ID) ([]txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], int, error)
+	UnconfirmedTransactions(limit, offset int, fromAddress ADDR, chainID CHAIN_ID) ([]txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], int, error)
+	ConfirmedTransactions(limit, offset int, fromAddress ADDR, chainID CHAIN_ID) ([]txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], int, error)
+	ConfirmedMissingReceiptTransactions(limit, offset int, fromAddress ADDR, chainID CHAIN_ID) ([]txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], int, error)
```

Comment:
Since these methods are reading from the DB, should we also have a context parameter? In case a close signal comes through and we have to cancel the context?